src/qt3utils/datagenerators/plescanner.py

```diff
@@ -0,0 +1,266 @@
+import logging
+import time
+import tkinter as tk
+from typing import Type
+
+import numpy as np
+
+from qt3utils.applications.controllers.nidaqedgecounter import QT3ScanNIDAQEdgeCounterController
+from qt3utils.applications.controllers.wavemeter_controller import WavemeterController
+from qt3utils.applications.controllers.lockin_controller import Lockin
+from qt3utils.nidaq.customcontrollers import VControl
+
+logger = logging.getLogger(__name__)
+
+class PleScanner:
+    """
+    Base class with function methods which any PLE experiment will need to operate
+    Override this class for specific types of PLE experiments
+    """
+
+    def __init__(self, readers, wavelength_controller):
+        """
+        self.running: boolean true when scan thread is running
+        self.current_frame: current pixel (or time) number in sweep
+        self._start: minimum allowed point for hardware controlling the laser wavelength
+        self._end: maximum allowed point for hardware controlling the laser wavelength
+        self.tmax: maximum number of pixels (or time) for sweep
+        self._step_size: size between steps controlling wavelength as it is swept
+        self.wavelength_controller: the wavelength controller class instance
+        """
```

Comment:
To properly declare this in the docstring, you put it in the class docstring as attributes:

```python
"""
Attributes
----------
running: bool
    true when scan thread is running
current_frame: int | float
    current pixel (or time) number in sweep
...
"""
```